src/messaging/ChannelContext.h

```diff
@@ -0,0 +1,199 @@
+/*
+ *
+ *    Copyright (c) 2020 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *    @file
+ *      This file defines the internal classes used by CHIP Channel.
+ */
+
+#pragma once
+
+#include <variant>
+
+#include <lib/core/ReferenceCounted.h>
+#include <lib/mdns/platform/Mdns.h>
+#include <messaging/Channel.h>
+#include <transport/CASESession.h>
+#include <transport/PeerConnectionState.h>
+#include <transport/SecureSessionMgr.h>
+
+namespace chip {
+namespace Messaging {
+
+class ExchangeManager;
+class ChannelContext;
+
+class ChannelContextDeletor
+{
+public:
+    static void Release(ChannelContext * context);
+};
+
+/**
+ * @brief
+ *  The object of the class holds all state of a channel. It is a state machine, with following states:
+ *
+ *  N: None, the initial state
+ *  P: Preparing
+ *  R: Ready, the channel is ready to use
+ *  C: Closed, the channel is closed
+ *  F: Failed, the channel is failed
+ *
+ *    +---+   +---+   +---+   +---+
+ *    | N |-->| P |-->| R |-->| C |
+ *    +---+   +---+   +---+   +---+
+ *              |       |
+ *              |       |     +---+
+ *              +-------+---->| F |
+ *                            +---+
+ *
+ *  Note: The state never goes back, when a channel is failed, it can't be reset or fixed. The application must create a
+ *  new channel to replace the failed channel
+ *
+ *  Preparing Substates:
+ *    A: AddressResolving, use mDNS to resolve the node address
+ *    P: PasePairing, do SPAKE2 key exchange
```

Comment:
Can we update this and the diagram to remove PASE related states?

Reply:
done